apps/meteor/ee/server/services/mongo.ts

```diff
@@ -14,23 +15,40 @@ export enum Collections {
 	Settings = 'rocketchat_settings',
 }
 
-let db: Db;
-export async function getConnection(poolSize = 5): Promise<Db> {
-	if (!db) {
-		const client = new MongoClient(MONGO_URL, {
-			useUnifiedTopology: true,
-			useNewUrlParser: true,
-			poolSize,
-		});
+function connectDb(poolSize: number): Promise<MongoClient> {
+	const client = new MongoClient(MONGO_URL, {
+		useUnifiedTopology: true,
+		useNewUrlParser: true,
+		poolSize,
+	});
 
-		await client.connect();
-		db = client.db(name);
-	}
-
-	return db;
+	return client.connect().catch((error) => {
+		// exits the process in case of any error
+		console.error(error);
+		process.exit(1);
+	});
 }
 
+export const getConnection = ((): ((poolSize?: number) => Promise<Db>) => {
+	let db: Db;
+	let client: Promise<MongoClient>;
+
+	return async (poolSize = 5): Promise<Db> => {
+		if (db) {
+			return db;
+		}
+		if (!client) {
+			client = connectDb(poolSize);
+			client.then((c) => {
+				db = c.db(name);
+			});
+		}
+		// if getConnection was called multiple times before it was connected, wait for the connection
+		return (await client).db(name);
+	};
+})();
+
 export async function getCollection<T>(name: Collections): Promise<Collection<T>> {
```

Comment:
i don't know how often we call `getCollection`, but if it's often, maybe we can cache `db` and call getConnection again only if db doesn't have a value instead of evry time we call getCollection 👀

Reply:
Yup, but it means you're spinning up a promise that won't be needed if db is already set 👀

Reply:
Something like this:
```
let db: Db;

async function getConnection() => client.connect().asfasdfdsaf
async function getCollection() => {
  if (!db)
      db = await getConnection();
   
   return db.collection('blablabla')
}
```

This way, the getConnection promise is only called (hopefully) once and getCollection manages the fetching of it, avoiding to call the extra promise